clang/lib/Sema/SemaOverload.cpp

```diff
@@ -6926,11 +6926,26 @@ void Sema::AddOverloadCandidate(
     /// have linkage. So that all entities of the same should share one
     /// linkage. But in clang, different entities of the same could have
     /// different linkage.
-    NamedDecl *ND = Function;
-    if (auto *SpecInfo = Function->getTemplateSpecializationInfo())
+    const NamedDecl *ND = Function;
+    bool IsImplicitlyInstantiated = false;
+    if (auto *SpecInfo = Function->getTemplateSpecializationInfo()) {
       ND = SpecInfo->getTemplate();
-
-    if (ND->getFormalLinkage() == Linkage::Internal) {
+      IsImplicitlyInstantiated = SpecInfo->getTemplateSpecializationKind() ==
+                                 TSK_ImplicitInstantiation;
+    }
+
+    /// Don't remove inline functions with internal linkage from the overload
+    /// set if they are declared in a GMF.
+    /// The global module is meant to be a transition mechanism for C and C++
+    /// headers.
+    /// Inline functions with internal linkage are a common pattern in headers
+    /// to avoid ODR issues.
+    const bool IsInlineFunctionInGMF =
+        Function->getOwningModule() &&
+        Function->getOwningModule()->isGlobalModule() &&
+        (IsImplicitlyInstantiated || Function->isInlined());
```

Comment:
Maybe we can replace `(IsImplicitlyInstantiated || Function->isInlined())` with `getASTContext().GetGVALinkageForFunction(Function) == GVA_DiscardableODR`.

Reply:
The GVA linkage seems to be sometimes `GVA_Internal` for some of the relevant functions, when running the new tests:

```c++
    if (IsImplicitlyInstantiated) {
      std::cerr << "IsImplicitlyInstantiated: GVA linkage: " << getASTContext().GetGVALinkageForFunction(Function) << '\n';
    }
    if (Function->isInlined()) {
      std::cerr << "Function->isInlined():    GVA linkage: " << getASTContext().GetGVALinkageForFunction(Function) << '\n';
    }
```

```
IsImplicitlyInstantiated: GVA linkage: 2
IsImplicitlyInstantiated: GVA linkage: 0
IsImplicitlyInstantiated: GVA linkage: 2
IsImplicitlyInstantiated: GVA linkage: 0
IsImplicitlyInstantiated: GVA linkage: 2
Function->isInlined():    GVA linkage: 0
IsImplicitlyInstantiated: GVA linkage: 2
Function->isInlined():    GVA linkage: 0
```

"0" is `GVA_Internal` and "2" is `GVA_DiscardableODR`. So for example `static inline` is `GVA_Internal`. So this wouldn't work I guess...

Reply:
Yeah... this was the reason why it was rejected..